### PR TITLE
fix meshblock pack issues coming from template shadowing and add a test

### LINKED ISSUE #570: Sparse interface update broke meshblockpacks
The following code snippet now fails
```C++
/ /MeshBlockData *rc provided above somewhere
PackIndexMap imap;
rc->PackVariables(names, imap);
```
with errors that look like
```bash
/parthenon/src/interface/mesh_data.hpp:296:5: error: no matching function for call to ‘parthenon::MeshBlockData<double>::PackVariables(const std::vector<std::__cxx11::basic_string<char> >&, parthenon::PackIndexMap&, parthenon::PackIndexMap&, std::vector<std::__cxx11::basic_string<char> >&)’
```
This function signature is obviously wrong. Somehow the `PackIndexMap` is getting passed in twice. I believe this is because the API

We didn't catch this because for `MeshBlockPack` we apparently don't test this particular call signature anywhere. This needs to change, as it is used everywhere in the downstream codes.

### PR DESCRIPTION
**Pull request by Yurlungur**
## PR Summary

Downstream, @brryan noticed that calls to `MeshData::PackVariables` with a `PackIndexMap`, for example
```C++
// MeshData *rc defined above
PackIndexMap imap;
rc->PackVariables(names, imap);
```
cause compile-time errors. The issue was introduced in the sparse interface update, #535 , where this machinery changed significantly. In short, in that update, explicit overloads were replaced with variatic templates. Unfortunately, C++ does not resolve these templates appropriately, and they shadow each other, causing bugs. We didn't catch this until now because templated functions are optimized out if they're not called.

This MR resolves #570 . I replace the variatics with explicit overloads again, with a few changes to reduce line number as best I can. I also add a call to this machinery in the calculate pi example, so this issue should be caught by regression tests in the future.

Note that the full regression test suite is currently failing. (See #571 ). So we should either merge this in without triggering the full suite on the CI, or we should wait for #571 to be resolved. I lean towards the former. Thoughts? 

## PR Checklist

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [x] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

File: src/interface/mesh_data.hpp [hunk 1]
```diff
@@ -289,50 +289,211 @@ class MeshData {
...
+                                     vpack_types::StringPair &key) {
+    return PackVariablesAndFluxesImpl(nullptr, &key, sparse_ids);
+  }
+  const auto &PackVariablesAndFluxes(const std::vector<int> &sparse_ids,
+                                     PackIndexMap &map) {
+    return PackVariablesAndFluxesImpl(&map, nullptr, sparse_ids);
+  }
+  const auto &PackVariablesAndFluxes(const std::vector<int> &sparse_ids) {
+    return PackVariablesAndFluxesImpl(nullptr, nullptr, sparse_ids);
+  }
```
Comment (pgrete) [inline src/interface/mesh_data.hpp:407]: Are these actually required?
If I'm not missing something they're matching the
```c++
  const auto &PackVariablesAndFluxes(const std::vector<Elem> &names_or_flags,
                                     PackIndexMap &map, vpack_types::StringPair &key)
```
above, aren't they?

**Review by pgrete — APPROVED** [review]
Pending the minor question, looks good! Thanks.
I'm also fine with merging this in while #571 is not fixed yet, but we should still run the full test suite here manually and see if nothing else new fails.

File: src/interface/mesh_data.hpp (hunk 1 above)
Comment (Yurlungur) [reply src/interface/mesh_data.hpp:407]: These exclude either the `PackIndexMap` or the key, or both. Whereas the one you quoted has both of them. They're just syntactic sugar overloads.

**Comment by Yurlungur** [comment]
Thanks @pgrete . Agreed, I'll run the full test suite

File: src/interface/mesh_data.hpp [hunk 2]
```diff
@@ -289,50 +289,211 @@ class MeshData {
...
+  const auto &PackVariablesAndFluxes(vpack_types::StringPair &key) {
+    return PackVariablesAndFluxesImpl(nullptr, &key);
+  }
+  const auto &PackVariablesAndFluxes(PackIndexMap &map) {
+    return PackVariablesAndFluxesImpl(&map, nullptr);
+  }
+  const auto &PackVariablesAndFluxes() {
+    return PackVariablesAndFluxesImpl(nullptr, nullptr);
+  }
```
Comment (brryan) [inline src/interface/mesh_data.hpp:420]: What does this return with no arguments? (For some reason I can't find the two-argument version of `PackVariablesAndFluxesImpl` to look at the code itself)

**Review by brryan — APPROVED** [review]
Just one question for my own edification. Looks good, thanks for adding a typical use case to the `calculate_pi` example.

File: src/interface/mesh_data.hpp (hunk 2 above)
Comment (Yurlungur) [reply src/interface/mesh_data.hpp:420]: It should pack all variables and all fluxes.

**Comment by Yurlungur** [comment]
> Pending the minor question, looks good! Thanks.
> I'm also fine with merging this in while #571 is not fixed yet, but we should still run the full test suite here manually and see if nothing else new fails.

Looks like the only error is the `Input` metadata diff

**Pull request merged**